runner/runner.go

```diff
@@ -231,59 +240,85 @@ func (r *Runner) run(ctx context.Context, checkMap map[string]*Check, list bool,
 		}
 	}
 
+	errs := make(chan responseError, len(currentCheckList))
+	responses := make(chan *Response, len(currentCheckList))
+
 	// main loop to get the checks info.
 	for _, name := range currentCheckList {
-		resp := &Response{
-			name: name,
-		}
 
-		currentCheck, ok := checkMap[name]
-		if !ok {
-			combinedResponse.errs["check not found"] = resp
-			combinedResponse.checkNotFound = true
-			continue
-		}
+		go func(name string) {
+			currentCheck, ok := checkMap[name]
+			if !ok {
+				errs <- responseError{"Check not found", name, true, nil}
+				return
+			}
 
-		// find runner for the given role only.
-		if !currentCheck.verifyRole(r.role) {
-			continue
-		}
+			// find runner for the given role only.
+			if !currentCheck.verifyRole(r.role) {
+				responses <- nil
+				return
+			}
 
-		var (
-			combinedOutput []byte
-			code           int
-			err            error
-			checkDuration  string
-		)
+			if _, ok := combinedResponse.checks[name]; ok {
+				errs <- responseError{"Duplicate check", name, false, nil}
+				return
+			}
 
-		if _, ok := combinedResponse.checks[name]; ok {
-			combinedResponse.errs["duplicate check"] = resp
-			continue
-		}
+			resp := &Response{
+				name: name,
+			}
 
-		// list option disables the check execution
-		if !list {
+			var (
+				combinedOutput []byte
+				code           int
+				err            error
+				checkDuration  string
+			)
 
-			start := time.Now()
-			combinedOutput, code, err = currentCheck.Run(ctx, r.role)
-			checkDuration = time.Since(start).String()
-		}
 
-		resp.output = string(combinedOutput)
-		resp.status = code
-		resp.duration = checkDuration
-		resp.description = currentCheck.Description
-		resp.cmd = currentCheck.Cmd
-		resp.timeout = currentCheck.Timeout
-		resp.list = list
-		combinedResponse.checks[name] = resp
-
-		// collect errors
-		if err != nil {
-			combinedResponse.errs[err.Error()] = resp
-		}
+			// list option disables the check execution
+			if !list {
+				start := time.Now()
+				combinedOutput, code, err = currentCheck.Run(ctx, r.role)
+				checkDuration = time.Since(start).String()
+			}
+
+			resp.output = string(combinedOutput)
+			resp.status = code
+			resp.duration = checkDuration
+			resp.description = currentCheck.Description
+			resp.cmd = currentCheck.Cmd
+			resp.timeout = currentCheck.Timeout
+			resp.list = list
+
+			responses <- resp
 
-		combinedResponse.status = max(combinedResponse.status, code)
+			// collect errors
+			if err != nil {
+				errs <- responseError{err.Error(), name, false, resp}
+			}
+		}(name)
+	}
+
+	for range currentCheckList {
+		select {
+		case err := <- errs:
```

Comment:
if we rename `errs` to `results` and remove `responses` in favour of using `resultError.response` then we can merge the two `case` statements here.